src/main/java/com/gmail/inverseconduit/bot/DefaultBot.java

```diff
@@ -53,14 +53,12 @@ public static DefaultBot create(ChatInterface chatInterface) {
 
     @Override
     public void start() {
-        executor.scheduleAtFixedRate(this::processMessageQueue, 1, 500, TimeUnit.MILLISECONDS);
+        // asynchronously enqueue the processing by blocking supplier
+        processingThread.submit(this::processMessageQueue);
     }
 
     private void processMessageQueue() {
-        while (messageQueue.peek() != null) {
-            LOGGER.finest("processing message from queue");
-            processingThread.submit(() -> processMessage(messageQueue.poll()));
-        }
+        Stream.generate(blockingMessageSupplier).forEach(headMessage -> processingThread.submit(() -> processMessage(headMessage)));
```

Comment:
Reverting to a classic for loop means losing the benefit of the blockingMessageSupplier and adding yet another Thread Pool that needs to be taken care of.

While I agree that a classic loop may be more readable / conventional I personally think that the blocking wait is cleaner